File: email-backup.java/src/test/java/demo/MimeParser.java
package demo;

import com.google.common.io.ByteStreams;

import javax.mail.Message;
import javax.mail.MessagingException;
import javax.mail.Multipart;
import javax.mail.Part;
import javax.mail.internet.MimeBodyPart;
import javax.mail.internet.MimeMessage;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class MimeParser {

    public static Result parse(MimeMessage message) throws IOException, MessagingException {
        return new Parser().parse(message);
    }

    static class Result {

        private final String content;
        private final Map<String, byte[]> files;

        public Result(String content, Map<String, byte[]> files) {
            this.content = content;
            this.files = files;
        }

        public String getContent() {
            return content;
        }

        public Map<String, byte[]> getFiles() {
            return files;
        }
    }

    private static class Parser {

        private String content;
        private Map<String, byte[]> files = new HashMap<>();

        private Result parse(Message message) throws IOException, MessagingException {
            String contentType = message.getContentType();

            if (contentType.contains("multipart")) {
                Multipart multipart = (Multipart) message.getContent();
                for (int i = 0; i < multipart.getCount(); i++) {
                    MimeBodyPart part = (MimeBodyPart) multipart.getBodyPart(i);
                    if (Part.ATTACHMENT.equalsIgnoreCase(part.getDisposition())) {
                        files.put(part.getFileName(), ByteStreams.toByteArray(part.getInputStream()));
                    } else {
                        content = part.getContent().toString();
                    }
                }
            } else if (contentType.contains("text/plain") || contentType.contains("text/html")) {
                Object content = message.getContent();
                if (content != null) {
                    this.content = content.toString();
                }
            }

            return new Result(content, files);
        }
    }
}

File: email-backup.kotlin/src/test/kotlin/demo/MimeParser.kt
package demo

import com.google.common.io.ByteStreams
import java.io.IOException
import java.util.HashMap
import javax.mail.Message
import javax.mail.MessagingException
import javax.mail.Multipart
import javax.mail.Part
import javax.mail.internet.MimeBodyPart
import javax.mail.internet.MimeMessage

public object MimeParser {

    @Throws(IOException::class, MessagingException::class)
    public fun parse(message: MimeMessage): Result {
        return Parser().parse(message)
    }

    class Result(public val content: String?, public val files: Map<String, ByteArray>)

    private class Parser {

        private var content: String? = null
        private val files = HashMap<String, ByteArray>()

        @Throws(IOException::class, MessagingException::class)
        fun parse(message: Message): Result {
            val contentType = message.contentType

            if (contentType.contains("multipart")) {
                val multipart = message.content as Multipart
                for (i in 0..multipart.count - 1) {
                    val part = multipart.getBodyPart(i) as MimeBodyPart
                    if (Part.ATTACHMENT.equals(part.disposition, ignoreCase = true)) {
                        files.put(part.fileName, ByteStreams.toByteArray(part.inputStream))
                    } else {
                        content = part.content.toString()
                    }
                }
            } else if (contentType.contains("text/plain") || contentType.contains("text/html")) {
                val content = message.content
                if (content != null) {
                    this.content = content.toString()
                }
            }

            return Result(content, files)
        }
    }
}